File: WEEK4/bj1_18352.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m, k, x] = input[0].split(' ').map(Number);
const adjList = Array.from({ length: n + 1 }, () => []);

for (let i = 1; i <= m; i++) {
    const [u, v] = input[i].split(' ').map(Number);
    adjList[u].push(v);
}

const distance = new Array(n + 1).fill(-1);
const queue = [];
let head = 0;

queue.push(x);
distance[x] = 0;

while (head < queue.length) {
    const current = queue[head++];

    for (const neighbor of adjList[current]) {
        if (distance[neighbor] === -1) {
            distance[neighbor] = distance[current] + 1;
            queue.push(neighbor);
        }
    }
}

const result = [];
for (let i = 1; i <= n; i++) {
    if (distance[i] === k) {
        result.push(i);
    }
}

if (result.length === 0) {
    console.log(-1);
} else {
    console.log(result.join('\n'));
}

File: WEEK4/bj1_18352.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class bj1_18352 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        int x = Integer.parseInt(st.nextToken());

        ArrayList<Integer>[] adjList = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) {
            adjList[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            adjList[u].add(v);
        }

        int[] distance = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            distance[i] = -1;
        }

        Queue<Integer> queue = new LinkedList<>();
        queue.add(x);
        distance[x] = 0;

        while (!queue.isEmpty()) {
            int current = queue.poll();

            for (int neighbor : adjList[current]) {
                if (distance[neighbor] == -1) {
                    distance[neighbor] = distance[current] + 1;
                    queue.add(neighbor);
                }
            }
        }

        ArrayList<Integer> result = new ArrayList<>();
        for (int i = 1; i <= n; i++) {
            if (distance[i] == k) {
                result.add(i);
            }
        }

        if (result.isEmpty()) {
            System.out.println(-1);
        } else {
            Collections.sort(result);
            StringBuilder sb = new StringBuilder();
            for (int city : result) {
                sb.append(city).append("\n");
            }
            System.out.print(sb);
        }
    }
}
